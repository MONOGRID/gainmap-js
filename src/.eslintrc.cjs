/** @type {import('eslint').ESLint.ConfigData} */
module.exports = {
  root: true,
  parser: '@typescript-eslint/parser',
  parserOptions: {
    project: true,
    tsconfigRootDir: __dirname
  },
  env: {
    browser: true,
    es2018: true
  },

  extends: [
    'plugin:@typescript-eslint/recommended',
    'plugin:@typescript-eslint/recommended-type-checked',
    'plugin:compat/recommended',
    'standard'
  ],

  plugins: [
    '@typescript-eslint',
    'simple-import-sort',
    'unused-imports'
  ],

  rules: {
    "@typescript-eslint/no-misused-promises": ["error", { checksVoidReturn: false }],
    "@typescript-eslint/no-duplicate-type-constituents": "off",
    // https://github.com/lydell/eslint-plugin-simple-import-sort/#usage
    'simple-import-sort/imports': 'error',
    'simple-import-sort/exports': 'error',
    // https://github.com/sweepline/eslint-plugin-unused-imports
    'no-unused-vars': 'off',
    '@typescript-eslint/no-unused-vars': 'off',
    'unused-imports/no-unused-imports': 'error',
    'unused-imports/no-unused-vars': ['warn', { vars: 'all', varsIgnorePattern: '^_', args: 'after-used', argsIgnorePattern: '^_' }]
  }
}
