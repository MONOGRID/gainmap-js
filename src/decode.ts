import {
  HalfFloatType,
  NoColorSpace,
  SRGBColorSpace
} from 'three'

import { GainMapDecoderMaterial } from './materials/GainMapDecoderMaterial'
import { DecodeParameters } from './types'
import { QuadRenderer } from './utils/QuadRenderer'

/**
 * Decodes a gainmap using a WebGLRenderTarget
 *
 * @category Decoding
 * @group Decoding
 * @example
 * import { decode } from 'gainmap-js'
 * import { ImageBitmapLoader, Mesh, PlaneGeometry, MeshBasicMaterial } from 'three'
 *
 * const loader = new ImageBitmapLoader()
 * loader.setOptions( { imageOrientation: 'flipY' } )
 *
 * // load SDR Representation
 * const sdr = await loader.loadAsync('sdr.jpg')
 * // load Gainmap recovery image
 * const gainMap = await loader.loadAsync('gainmap.jpg')
 * // load metadata
 * const metadata = await (await fetch('metadata.json')).json()
 *
 * const result = await decodeToRenderTarget({
 *   sdr,
 *   gainMap,
 *   // this will restore the full HDR range
 *   maxDisplayBoost: Math.pow(2, metadata.hdrCapacityMax)
 *   ...metadata,
 * })
 *
 * // result can be used to populate a Texture
 * const mesh = new Mesh(new PlaneGeometry(), new MeshBasicMaterial({ map: result.renderTarget.texture }))
 *
 * @param params
 * @returns
 * @throws {Error} if the WebGLRenderer fails to render the gainmap
 */
export const decode = (params: DecodeParameters): InstanceType<typeof QuadRenderer<typeof HalfFloatType, InstanceType<typeof GainMapDecoderMaterial>>> => {
  const { sdr, gainMap, renderer } = params

  if (sdr.colorSpace !== SRGBColorSpace) {
    console.warn('SDR Colorspace needs to be *SRGBColorSpace*, setting it automatically')
    sdr.colorSpace = SRGBColorSpace
  }
  sdr.needsUpdate = true

  if (gainMap.colorSpace !== NoColorSpace) {
    console.warn('Gainmap Colorspace needs to be *NoColorSpace*')
    gainMap.colorSpace = NoColorSpace
  }
  gainMap.needsUpdate = true

  const material = new GainMapDecoderMaterial({
    ...params,
    sdr,
    gainMap
  })

  const quadRenderer = new QuadRenderer(sdr.image.width, sdr.image.height, HalfFloatType, NoColorSpace, material, renderer)
  try {
    quadRenderer.render()
  } catch (e) {
    quadRenderer.dispose()
    throw e
  }
  return quadRenderer
}
