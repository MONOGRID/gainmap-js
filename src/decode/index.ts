export * from '../core/QuadRenderer'
export * from '../core/types'
export * from './decode'
export * from './extract'
export * from './loaders/GainMapLoader'
export * from './loaders/HDRJPGLoader'
// Legacy name, TODO: can be removed with next breaking change release
export { HDRJPGLoader as JPEGRLoader } from './loaders/HDRJPGLoader'
export * from './materials/GainMapDecoderMaterial'
export * from './types'
export * from './utils/extractXMP'
export * from './utils/MPFExtractor'
