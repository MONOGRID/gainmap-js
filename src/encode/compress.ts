import { CompressedImage, CompressParameters } from './types'

/**
 * Used internally
 *
 * @internal
 * @param canvas
 * @param mimeType
 * @param quality
 * @returns
 */
const canvasToBlob = async (canvas: OffscreenCanvas | HTMLCanvasElement, mimeType: CompressParameters['mimeType'], quality: CompressParameters['quality']) => {
  if (typeof OffscreenCanvas !== 'undefined' && canvas instanceof OffscreenCanvas) {
    return canvas.convertToBlob({ type: mimeType, quality: quality || 0.9 })
  } else if (canvas instanceof HTMLCanvasElement) {
    return new Promise<Blob>((resolve, reject) => {
      canvas.toBlob((res) => {
        if (res) resolve(res)
        else reject(new Error('Failed to convert canvas to blob'))
      }, mimeType, quality || 0.9)
    })
  }
  /* istanbul ignore next
    as long as this function is not exported this is only here
    to satisfy TS strict mode internally
  */
  throw new Error('Unsupported canvas element')
}

/**
 * Converts a RAW RGBA image buffer into the provided `mimeType` using the provided `quality`
 *
 * @category Compression
 * @group Compression
 * @param params
 * @throws {Error} if the browser does not support [createImageBitmap](https://caniuse.com/createimagebitmap)
 * @throws {Error} if the provided source image cannot be decoded
 * @throws {Error} if the function fails to create a canvas context
 */
export const compress = async (params: CompressParameters): Promise<CompressedImage> => {
  if (typeof createImageBitmap === 'undefined') throw new Error('createImageBitmap() not supported.')
  const { source, mimeType, quality, flipY } = params
  // eslint-disable-next-line no-undef
  let imageBitmapSource: ImageBitmapSource
  if ((source instanceof Uint8Array || source instanceof Uint8ClampedArray) && 'sourceMimeType' in params) {
    imageBitmapSource = new Blob([source], { type: params.sourceMimeType })
  } else if (source instanceof ImageData) {
    imageBitmapSource = source
  } else {
    throw new Error('Invalid source image')
  }
  const img = await createImageBitmap(imageBitmapSource)
  const width = img.width
  const height = img.height

  let canvas: OffscreenCanvas | HTMLCanvasElement
  if (typeof OffscreenCanvas !== 'undefined') {
    canvas = new OffscreenCanvas(width, height)
  } else {
    canvas = document.createElement('canvas')
    canvas.width = width
    canvas.height = height
  }
  const ctx = canvas.getContext('2d')
  if (!ctx) throw new Error('Failed to create canvas Context')
  // flip Y
  if (flipY === true) {
    ctx.translate(0, height)
    ctx.scale(1, -1)
  }

  ctx.drawImage(img, 0, 0, width, height)

  const blob = await canvasToBlob(canvas, mimeType, quality || 0.9)

  const data = new Uint8Array(await blob.arrayBuffer())

  return {
    data,
    mimeType,
    width,
    height
  }
}
