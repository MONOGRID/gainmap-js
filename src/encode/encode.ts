import { getDataTexture } from '../core/get-data-texture'
import { GainMapMetadata } from '../core/types'
import { getGainMap } from './get-gainmap'
import { getSDRRendition } from './get-sdr-rendition'
import { EncodingParametersBase } from './types'

/**
 * Encodes a Gainmap starting from an HDR file.
 *
 * @remarks
 * if you do not pass a `renderer` parameter
 * you must manually dispose the result
 * ```js
 * const encodingResult = await encode({ ... })
 * // do something with the buffers
 * const sdr = encodingResult.sdr.getArray()
 * const gainMap = encodingResult.gainMap.getArray()
 * // after that
 * encodingResult.sdr.dispose()
 * encodingResult.gainMap.dispose()
 * ```
 *
 * @category Encoding Functions
 * @group Encoding Functions
 *
 * @example
 * import { encode, findTextureMinMax } from '@monogrid/gainmap-js'
 * import { EXRLoader } from 'three/examples/jsm/loaders/EXRLoader.js'
 *
 * // load an HDR file
 * const loader = new EXRLoader()
 * const image = await loader.loadAsync('image.exr')
 *
 * // find RAW RGB Max value of a texture
 * const textureMax = await findTextureMinMax(image)
 *
 * // Encode the gainmap
 * const encodingResult = encode({
 *   image,
 *   // this will encode the full HDR range
 *   maxContentBoost: Math.max.apply(this, textureMax)
 * })
 * // can be re-encoded after changing parameters
 * encodingResult.sdr.material.exposure = 0.9
 * encodingResult.sdr.render()
 * // or
 * encodingResult.gainMap.material.gamma = [1.1, 1.1, 1.1]
 * encodingResult.gainMap.render()
 *
 * // do something with encodingResult.gainMap.toArray()
 * // and encodingResult.sdr.toArray()
 *
 * // renderers must be manually disposed
 * encodingResult.sdr.dispose()
 * encodingResult.gainMap.dispose()
 *
 * @param params Encoding Parameters
 * @returns
 */
export const encode = (params: EncodingParametersBase) => {
  const { image, renderer } = params

  const dataTexture = getDataTexture(image)

  const sdr = getSDRRendition(dataTexture, renderer, params.toneMapping, params.renderTargetOptions)

  const gainMapRenderer = getGainMap({
    ...params,
    image: dataTexture,
    sdr,
    renderer: sdr.renderer // reuse the same (maybe disposable?) renderer
  })

  return {
    sdr,
    gainMap: gainMapRenderer,
    hdr: dataTexture,
    getMetadata: (): GainMapMetadata => {
      return {
        gainMapMax: gainMapRenderer.material.gainMapMax,
        gainMapMin: gainMapRenderer.material.gainMapMin,
        gamma: gainMapRenderer.material.gamma,
        hdrCapacityMax: gainMapRenderer.material.hdrCapacityMax,
        hdrCapacityMin: gainMapRenderer.material.hdrCapacityMin,
        offsetHdr: gainMapRenderer.material.offsetHdr,
        offsetSdr: gainMapRenderer.material.offsetSdr
      }
    }
  }
}
