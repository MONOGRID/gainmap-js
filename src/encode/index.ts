export * from '../core/types'
export * from './compress'
export * from './encode'
export * from './encode-and-compress'
export * from './find-texture-min-max'
export * from './get-gainmap'
export * from './get-sdr-rendition'
export * from './materials/GainMapEncoderMaterial'
export * from './materials/SDRMaterial'
export * from './types'
