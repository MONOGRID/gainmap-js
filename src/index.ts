export * from './decode'
export * from './encode'
export * from './encode-utils/compress'
export * from './encode-utils/find-texture-min-max'
export * from './encode-utils/get-gainmap'
export * from './encode-utils/get-sdr-rendition'
export * from './materials/index'
export * from './materials/index'
export * from './types'
export * from './utils/get-data-texture'
export * from './utils/QuadRenderer'
