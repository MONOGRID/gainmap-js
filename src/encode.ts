import { compress } from './encode-utils/compress'
import { getGainMap } from './encode-utils/get-gainmap'
import { getSDRRendition } from './encode-utils/get-sdr-rendition'
import { CompressedImage, EncodingParametersBase, EncodingParametersWithCompression, GainMapMetadata } from './types'
import { getDataTexture } from './utils/get-data-texture'

/**
 * Encodes a Gainmap starting from an HDR file.
 *
 * Can optionally use a WebWorker to offload the encoding process to the worker
 *
 * @category Encoding Functions
 * @group Encoding Functions
 * @example
 * import { encode } from 'gainmap-js'
 * import { EXRLoader } from 'three/examples/jsm/loaders/EXRLoader'
 *
 * const loader = new EXRLoader()
 * const image = await loader.loadAsync('image.exr')
 * // This will:
 * // * create a WebGLRenderer
 * // * Render the Gainmap
 * // * dispose the WebGLRenderer
 *
 * const gainmap = await encode({image})
 *
 * @param params Encoding Paramaters
 * @returns
 */
export const encode = (params: EncodingParametersBase) => {
  const { image, renderer } = params

  const dataTexture = getDataTexture(image)

  const sdr = getSDRRendition(dataTexture, renderer, params.toneMapping)

  const gainMapRenderer = getGainMap({
    ...params,
    image: dataTexture,
    sdr,
    renderer: sdr.renderer // reuse the same (maybe disposable?) renderer
  })

  return {
    sdr,
    gainMap: gainMapRenderer,
    hdr: dataTexture,
    getMetadata: () => {
      const meta: GainMapMetadata = {
        gainMapMax: gainMapRenderer.material.gainMapMax,
        gainMapMin: gainMapRenderer.material.gainMapMin,
        gamma: gainMapRenderer.material.gamma,
        hdrCapacityMax: gainMapRenderer.material.hdrCapacityMax,
        hdrCapacityMin: gainMapRenderer.material.hdrCapacityMin,
        offsetHdr: gainMapRenderer.material.offsetHdr,
        offsetSdr: gainMapRenderer.material.offsetSdr
      }
      return meta
    }
  }
}

/**
 * Encodes a Gainmap starting from an HDR file into compressed file formats (`image/jpeg`, `image/webp` or `image/png`).
 *
 * Uses {@link encode} internally, then pipes the results to {@link compress}.
 *
 * @category Encoding Functions
 * @group Encoding Functions
 * @example
 *
 * @param params Encoding Paramaters
 * @throws {Error} if the browser does not support [createImageBitmap](https://caniuse.com/createimagebitmap)
 */
export const encodeAndCompress = async (params: EncodingParametersWithCompression) => {
  const encodingResult = encode(params)

  const { mimeType, quality, flipY, withWorker } = params

  let compressResult: [CompressedImage, CompressedImage]

  let rawSDR: Uint8ClampedArray
  let rawGainMap: Uint8ClampedArray

  const sdrImageData = new ImageData(encodingResult.sdr.toArray(), encodingResult.sdr.width, encodingResult.sdr.height)
  const gainMapImageData = new ImageData(encodingResult.gainMap.toArray(), encodingResult.gainMap.width, encodingResult.gainMap.height)

  if (withWorker) {
    const workerResult = await Promise.all([
      withWorker.compress({
        source: sdrImageData,
        mimeType,
        quality,
        flipY
      }),
      withWorker.compress({
        source: gainMapImageData,
        mimeType,
        quality,
        flipY
      })
    ])
    compressResult = workerResult
    rawSDR = workerResult[0].source
    rawGainMap = workerResult[1].source
  } else {
    compressResult = await Promise.all([
      compress({
        source: sdrImageData,
        mimeType,
        quality,
        flipY
      }),
      compress({
        source: gainMapImageData,
        mimeType,
        quality,
        flipY
      })
    ])
    rawSDR = sdrImageData.data
    rawGainMap = gainMapImageData.data
  }

  encodingResult.sdr.dispose()
  encodingResult.gainMap.dispose()

  return {
    ...encodingResult,
    ...encodingResult.getMetadata(),
    sdr: compressResult[0],
    gainMap: compressResult[1],
    rawSDR,
    rawGainMap
  }
}
