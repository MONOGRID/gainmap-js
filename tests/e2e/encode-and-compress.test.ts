import 'jest-extended'

import { describe, expect, it } from '@jest/globals'
import { toMatchImageSnapshot } from 'jest-image-snapshot'
import sharp from 'sharp'
import { ACESFilmicToneMapping, LinearToneMapping } from 'three'

import { encodeAndCompress } from '../../src/encode'
import { getPage } from './common'

expect.extend({ toMatchImageSnapshot })

const matrix: [string, string, number, number][] = [
  ['memorial.exr', 'jpg', 0.7, ACESFilmicToneMapping],
  ['memorial.exr', 'webp', 0.7, ACESFilmicToneMapping],
  ['memorial.exr', 'png', 0.7, ACESFilmicToneMapping],

  ['memorial.hdr', 'jpg', 0.7, ACESFilmicToneMapping],
  ['memorial.hdr', 'webp', 0.7, ACESFilmicToneMapping],
  ['memorial.hdr', 'png', 0.7, ACESFilmicToneMapping],

  ['memorial.exr', 'jpg', 0.7, LinearToneMapping],
  ['memorial.exr', 'webp', 0.7, LinearToneMapping],
  ['memorial.exr', 'png', 0.7, LinearToneMapping],

  ['memorial.hdr', 'jpg', 0.7, LinearToneMapping],
  ['memorial.hdr', 'webp', 0.7, LinearToneMapping],
  ['memorial.hdr', 'png', 0.7, LinearToneMapping],

  ['spruit_sunrise_1k.hdr', 'jpg', 0.95, LinearToneMapping]
]

describe('encode-and-compress', () => {
  it.each(matrix)('encodes %p to %p using quality %p, tonemapping: %p', async (file, format, quality, tonemapping) => {
    const { page, pageError, pageLog } = await getPage('base')

    await page.addScriptTag({
      type: 'module',
      url: 'scripts/encode-and-compress.js'
    })

    const result = await page.evaluate(`
        encodeAndCompress(
          '${file}',
          'image/${format === 'jpg' ? 'jpeg' : format}',
          ${quality},
          ${tonemapping}
      )`) as Awaited<ReturnType<typeof encodeAndCompress>> & { textureMax: [number, number, number] }

    expect(pageError).not.toBeCalled()
    // expect no calls to page log except the one indicated
    expect(pageLog).not.toBeCalledWith(
      // eslint-disable-next-line @typescript-eslint/no-unsafe-call
      expect.not.stringMatching(/GPU stall due to ReadPixels/)
    )

    // we receive Arrays because puppeteer can't transfer Uint8Array data
    result.gainMap.data = Uint8Array.from(result.gainMap.data)
    result.sdr.data = Uint8Array.from(result.sdr.data)

    expect(result.gainMapMin).toBeArrayOfSize(3)
    expect(result.gainMapMax).toBeArrayOfSize(3)

    expect(result.hdrCapacityMin).toBeNumber()
    expect(result.hdrCapacityMax).toBeNumber()

    expect(result.hdrCapacityMin).toBeLessThan(result.hdrCapacityMax)

    expect(result.textureMax).toBeArrayOfSize(3)
    expect(result.textureMax).not.toContain(0)

    expect(result.gamma).toBeArrayOfSize(3)
    expect(result.gamma).toContain(1)

    expect(result.offsetHdr).toBeArrayOfSize(3)
    expect(result.offsetHdr).not.toContain(0)

    expect(result.offsetSdr).toBeArrayOfSize(3)
    expect(result.offsetSdr).not.toContain(0)

    expect(result.gainMap).toBeObject()
    expect(result.gainMap.data).toBeInstanceOf(Uint8Array)
    expect(result.gainMap.data.length).toBeGreaterThan(4)

    expect(result.gainMap.width).toBeNumber()
    expect(result.gainMap.width).not.toBeNaN()
    expect(result.gainMap.width).not.toBe(0)

    expect(result.gainMap.height).toBeNumber()
    expect(result.gainMap.height).not.toBeNaN()
    expect(result.gainMap.height).not.toBe(0)

    expect(result.sdr).toBeObject()
    expect(result.sdr.data).toBeInstanceOf(Uint8Array)
    expect(result.sdr.data.length).toBeGreaterThan(4)

    expect(result.sdr.width).toBeNumber()
    expect(result.sdr.height).not.toBeNaN()
    expect(result.sdr.width).not.toBe(0)

    expect(result.sdr.height).toBeNumber()
    expect(result.sdr.height).not.toBeNaN()
    expect(result.sdr.height).not.toBe(0)

    expect({
      offsetSdr: result.offsetSdr,
      offsetHdr: result.offsetHdr,
      gamma: result.gamma,
      hdrCapacityMin: result.hdrCapacityMin,
      hdrCapacityMax: result.hdrCapacityMax,
      gainMapMin: result.gainMapMin,
      gainMapMax: result.gainMapMax
    }).toMatchSnapshot('metadata')

    expect(
      await sharp(result.sdr.data)
        .resize({ width: 500, height: 500, fit: 'inside' })
        .png({ compressionLevel: 9, effort: 10 })
        .toBuffer()
    ).toMatchImageSnapshot({
      comparisonMethod: 'ssim',
      failureThreshold: 0.015, // 1.5% difference
      failureThresholdType: 'percent'
    })
    expect(
      await sharp(result.gainMap.data)
        .resize({ width: 500, height: 500, fit: 'inside' })
        .png({ compressionLevel: 9, effort: 10 })
        .toBuffer()
    ).toMatchImageSnapshot({
      comparisonMethod: 'ssim',
      failureThreshold: 0.015, // 1.5% difference
      failureThresholdType: 'percent'
    })

    await page.close()
  }, 900000 /* 15 minutes */)
})
