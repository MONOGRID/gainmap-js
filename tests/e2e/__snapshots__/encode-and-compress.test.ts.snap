// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`encode-and-compress encodes "memorial.exr" to "jpg" using quality 0.7, tonemapping: 1 1`] = `
{
  "gainMapMax": [
    8.098032082960527,
    8.098032082960527,
    8.098032082960527,
  ],
  "gainMapMin": [
    0,
    0,
    0,
  ],
  "gamma": [
    1,
    1,
    1,
  ],
  "hdrCapacityMax": 8.098032082960527,
  "hdrCapacityMin": 0,
  "offsetHdr": [
    0.015625,
    0.015625,
    0.015625,
  ],
  "offsetSdr": [
    0.015625,
    0.015625,
    0.015625,
  ],
}
`;

exports[`encode-and-compress encodes "memorial.exr" to "jpg" using quality 0.7, tonemapping: 4 1`] = `
{
  "gainMapMax": [
    8.098032082960527,
    8.098032082960527,
    8.098032082960527,
  ],
  "gainMapMin": [
    0,
    0,
    0,
  ],
  "gamma": [
    1,
    1,
    1,
  ],
  "hdrCapacityMax": 8.098032082960527,
  "hdrCapacityMin": 0,
  "offsetHdr": [
    0.015625,
    0.015625,
    0.015625,
  ],
  "offsetSdr": [
    0.015625,
    0.015625,
    0.015625,
  ],
}
`;

exports[`encode-and-compress encodes "memorial.exr" to "png" using quality 0.7, tonemapping: 1 1`] = `
{
  "gainMapMax": [
    8.098032082960527,
    8.098032082960527,
    8.098032082960527,
  ],
  "gainMapMin": [
    0,
    0,
    0,
  ],
  "gamma": [
    1,
    1,
    1,
  ],
  "hdrCapacityMax": 8.098032082960527,
  "hdrCapacityMin": 0,
  "offsetHdr": [
    0.015625,
    0.015625,
    0.015625,
  ],
  "offsetSdr": [
    0.015625,
    0.015625,
    0.015625,
  ],
}
`;

exports[`encode-and-compress encodes "memorial.exr" to "png" using quality 0.7, tonemapping: 4 1`] = `
{
  "gainMapMax": [
    8.098032082960527,
    8.098032082960527,
    8.098032082960527,
  ],
  "gainMapMin": [
    0,
    0,
    0,
  ],
  "gamma": [
    1,
    1,
    1,
  ],
  "hdrCapacityMax": 8.098032082960527,
  "hdrCapacityMin": 0,
  "offsetHdr": [
    0.015625,
    0.015625,
    0.015625,
  ],
  "offsetSdr": [
    0.015625,
    0.015625,
    0.015625,
  ],
}
`;

exports[`encode-and-compress encodes "memorial.exr" to "webp" using quality 0.7, tonemapping: 1 1`] = `
{
  "gainMapMax": [
    8.098032082960527,
    8.098032082960527,
    8.098032082960527,
  ],
  "gainMapMin": [
    0,
    0,
    0,
  ],
  "gamma": [
    1,
    1,
    1,
  ],
  "hdrCapacityMax": 8.098032082960527,
  "hdrCapacityMin": 0,
  "offsetHdr": [
    0.015625,
    0.015625,
    0.015625,
  ],
  "offsetSdr": [
    0.015625,
    0.015625,
    0.015625,
  ],
}
`;

exports[`encode-and-compress encodes "memorial.exr" to "webp" using quality 0.7, tonemapping: 4 1`] = `
{
  "gainMapMax": [
    8.098032082960527,
    8.098032082960527,
    8.098032082960527,
  ],
  "gainMapMin": [
    0,
    0,
    0,
  ],
  "gamma": [
    1,
    1,
    1,
  ],
  "hdrCapacityMax": 8.098032082960527,
  "hdrCapacityMin": 0,
  "offsetHdr": [
    0.015625,
    0.015625,
    0.015625,
  ],
  "offsetSdr": [
    0.015625,
    0.015625,
    0.015625,
  ],
}
`;

exports[`encode-and-compress encodes "memorial.hdr" to "jpg" using quality 0.7, tonemapping: 1 1`] = `
{
  "gainMapMax": [
    8.103287808412022,
    8.103287808412022,
    8.103287808412022,
  ],
  "gainMapMin": [
    0,
    0,
    0,
  ],
  "gamma": [
    1,
    1,
    1,
  ],
  "hdrCapacityMax": 8.103287808412022,
  "hdrCapacityMin": 0,
  "offsetHdr": [
    0.015625,
    0.015625,
    0.015625,
  ],
  "offsetSdr": [
    0.015625,
    0.015625,
    0.015625,
  ],
}
`;

exports[`encode-and-compress encodes "memorial.hdr" to "jpg" using quality 0.7, tonemapping: 4 1`] = `
{
  "gainMapMax": [
    8.103287808412022,
    8.103287808412022,
    8.103287808412022,
  ],
  "gainMapMin": [
    0,
    0,
    0,
  ],
  "gamma": [
    1,
    1,
    1,
  ],
  "hdrCapacityMax": 8.103287808412022,
  "hdrCapacityMin": 0,
  "offsetHdr": [
    0.015625,
    0.015625,
    0.015625,
  ],
  "offsetSdr": [
    0.015625,
    0.015625,
    0.015625,
  ],
}
`;

exports[`encode-and-compress encodes "memorial.hdr" to "png" using quality 0.7, tonemapping: 1 1`] = `
{
  "gainMapMax": [
    8.103287808412022,
    8.103287808412022,
    8.103287808412022,
  ],
  "gainMapMin": [
    0,
    0,
    0,
  ],
  "gamma": [
    1,
    1,
    1,
  ],
  "hdrCapacityMax": 8.103287808412022,
  "hdrCapacityMin": 0,
  "offsetHdr": [
    0.015625,
    0.015625,
    0.015625,
  ],
  "offsetSdr": [
    0.015625,
    0.015625,
    0.015625,
  ],
}
`;

exports[`encode-and-compress encodes "memorial.hdr" to "png" using quality 0.7, tonemapping: 4 1`] = `
{
  "gainMapMax": [
    8.103287808412022,
    8.103287808412022,
    8.103287808412022,
  ],
  "gainMapMin": [
    0,
    0,
    0,
  ],
  "gamma": [
    1,
    1,
    1,
  ],
  "hdrCapacityMax": 8.103287808412022,
  "hdrCapacityMin": 0,
  "offsetHdr": [
    0.015625,
    0.015625,
    0.015625,
  ],
  "offsetSdr": [
    0.015625,
    0.015625,
    0.015625,
  ],
}
`;

exports[`encode-and-compress encodes "memorial.hdr" to "webp" using quality 0.7, tonemapping: 1 1`] = `
{
  "gainMapMax": [
    8.103287808412022,
    8.103287808412022,
    8.103287808412022,
  ],
  "gainMapMin": [
    0,
    0,
    0,
  ],
  "gamma": [
    1,
    1,
    1,
  ],
  "hdrCapacityMax": 8.103287808412022,
  "hdrCapacityMin": 0,
  "offsetHdr": [
    0.015625,
    0.015625,
    0.015625,
  ],
  "offsetSdr": [
    0.015625,
    0.015625,
    0.015625,
  ],
}
`;

exports[`encode-and-compress encodes "memorial.hdr" to "webp" using quality 0.7, tonemapping: 4 1`] = `
{
  "gainMapMax": [
    8.103287808412022,
    8.103287808412022,
    8.103287808412022,
  ],
  "gainMapMin": [
    0,
    0,
    0,
  ],
  "gamma": [
    1,
    1,
    1,
  ],
  "hdrCapacityMax": 8.103287808412022,
  "hdrCapacityMin": 0,
  "offsetHdr": [
    0.015625,
    0.015625,
    0.015625,
  ],
  "offsetSdr": [
    0.015625,
    0.015625,
    0.015625,
  ],
}
`;

exports[`encode-and-compress encodes "spruit_sunrise_1k.hdr" to "jpg" using quality 0.95, tonemapping: 1 1`] = `
{
  "gainMapMax": [
    15.99929538702341,
    15.99929538702341,
    15.99929538702341,
  ],
  "gainMapMin": [
    0,
    0,
    0,
  ],
  "gamma": [
    1,
    1,
    1,
  ],
  "hdrCapacityMax": 15.99929538702341,
  "hdrCapacityMin": 0,
  "offsetHdr": [
    0.015625,
    0.015625,
    0.015625,
  ],
  "offsetSdr": [
    0.015625,
    0.015625,
    0.015625,
  ],
}
`;
